### PR TITLE
Allow the script to be sourced

### PR DESCRIPTION
**Pull request by tvillega**
Hi there!

This PR follows the [bpkg guidelines](https://bpkg.sh/guidelines/) to allow the script to be sourced and run the functions 'natively' inside the shell.

Nothing fancy, should do the job.

### CODE REVIEW

**Comment by ranger6** [comment]
Thanks for your interest and PR for the semver tool.  I'm pleased to see that the tool
can be re-purposed without much fuss: hints that we don't have cruft and unneccessary
dependencies hanging around.

I have several questions and comments:

1. What, exactly, is the problem being solved or advantage being provided?  I don't see any performance advantage.  The interface (parameters and return values) to `semver` remains the same: text parameters and text on stdout; errors on stderr.  Embedding the script (sourcing) eliminates the need to have the tool (correctly) found on the PATH: this can be an annoyance when `semver` needs to be run from within an automation tool (is this the motivation?).  However, it would seem that this is just kicking the can down the street: the enclosing script needs to be on the PATH *and* one has to make sure the script's location can be found by the sourcing script.  Can't figure it out.  This motivation should be in the PR submission.
2. My main concern is that the API is being widened.  First, the tool can be invoked (the current API), or sourced and invoked (the `if` statement implements the two interfaces).  Second, the user is being encouraged to incorporate the tool more as a library rather than a single, narrow call.  What happens when we change the implementation, names? We break people's code.  Also, some of the (as of now) internal functions make assumptions about the calling arguments: we can avoid checking and handling errors. Fortunately we don't have local state (except the few "constant" definitions like "PROG" and "DIGITS"), so the code is re-entrant (necessary if sourced).  What if that changes in the future? Once published as a library, all bets are off. 
3. The multiple "export" statements added advertise the wider interface: a well defined library. Not all the "internal" functions are exported: why? or why not?  If "export" is meant to solve a visibility problem when `semver_tool` is invoked, all the internal functions would need to be exported.  In fact, it seems to me that the "exports" serve no purpose.  When I wrapped a version of the script with the _exports removed_ with the following code:
```
#!/usr/bin/env bash

. ./semver

./semver "${@}"
semver_tool "${@}"
echo $(semver_tool "${@}")
```
I see no difference in behaviour between the three invocations.  Note that the "$( ... )" runs in a sub-shell.

4. I see that you renamed several functions (prefix with `semver_`).  Is this just preference, or, is it an attempt to avoid name collisions within the enclosing (sourcing) script?  If it is collisions, then the problem remains for all the other functions.  Worse, these collisions won't be caught by the shell: functions will silently get re-defined during or after the sourcing.  Here is one advantage of using the current, narrow interface: collisions are not possible. The same problem exists for variables: e.g. imagine that the definition of DIGITS changed!
5. You mention the "bpkg" standard.  I assume that you are not proposing making the semver tool a package with a json dependency file, etc.  You simply applied their (neat!) trick of implementing source/invoke in the same script.
6. Error reporting: at present, errors are reported from the perspective of a stand-alone script: e.g. we indicate usage, we claim that the program name is "semver", orient the user to the "-h" option, etc.  Also, `set` is invoked that would change global options for the enclosing script.
7. Tests: changes in functionality *must* have automated tests added.
8. Documentation: the user interface is pretty well documented, including examples.  The same would be needed if sourcing is supported (including caveats about names, errors, ..).
9. Commit message: if you look at our git log, you will see the sort of commit messages we've produced.  One can understand the evolution of the tool without reverse engineering the code.  "Allow the script to be sourced" is not enough.

In short, I don't think the PR is ready to go.  If the issues above can be addressed (especially the first, "why?"), then maybe something of reduced scope might work: ditch the exports and renaming, simply add the test for sourcing and replacing `exit` with `return`, add unit tests, and document: "_If you want to, you can source the script.  There are no compatibility guarantees going forward and watch out for the gotcha's._"

**Comment by tvillega** [comment]
>Thanks for you interest and PR for the semver tool.  I'm pleased to see that the tool
can be re-purposed without much fuss: hints that we don't have cruft and unneccessary
dependencies hanging around.

I'm glad to hear back from you and to know this repository is still receiving feedback.

>I have several questions and comments:

I'll try to answer to your points thoroughly.

>1. What, exactly, is the problem being solved or advantage being provided?  I don't see any performance advantage.  The interface (parameters and return values) to `semver` remains the same: text parameters and text on stdout; errors on stderr.  Embedding the script (sourcing) eliminates the need to have the tool (correctly) found on the PATH: this can be an annoyance when `semver` needs to be run from within an automation tool (is this the motivation?).  However, it would seem that this is just kicking the can down the street: the enclosing script needs to be on the PATH *and* one has to make sure the script's location can be found by the sourcing script.  Can't figure it out.  This motivation should be in the PR submission.

The problem being solved is the inability to source the `semver` script inside another script, because it was not designed to do so (it is not advertised as a feature). The advantages provided by this solution only apply to the use case where the script does not have execution permissions and it is only allowed to be embedded inside another script.

>2. My main concern is that the API is being widened.  First, the tool can be invoked (the current API), or sourced and invoked (the `if` statement implements the two interfaces).  Second, the user is being encouraged to incorporate the tool more as a library rather than a single, narrow call.  What happens when we change the implementation, names: we break people's code.  Also, some of the (as of now) internal functions make assumptions about the calling arguments: we can avoid checking and handling errors. Fortunately we don't have local state (except the few "constant" definitions like "PROG" and "DIGITS"), so the code is re-entrant (necessary if sourced).  What if that changes in the future? Once published as a library, all bets are off.

Yes, I widened the API by allowing the script to be sourced. I did not mean to remove or replace the current functionality, therefore any incompatibility that my PR introduces are bugs.

I am not against the statements you make, those are real issues I did not take into consideration. To keep my answers relevant to the points, I'll expand on my reasoning later.

>3. The multiple "export" statements added advertise the wider interface: a well defined library. Not all the "internal" functions are exported: why? or why not?  If "export" is meant to solve a visibility problem when `semver_tool` is invoked, all the internal functions would need to be exported.  In fact, it seems to me that the "exports" serve no purpose.  When I wrapped a version of the script with the _exports removed_ with the following code:

```
#!/usr/bin/env bash

. ./semver

semver "${@}"
semver_tool "${@}"
echo $(semver_tool "${@}")
```
>I so no difference in behaviour between the three invocations.  Note that the "$( ... )" runs in a sub-shell.

The internal functions that I exported are the ones inside the case switch. I understood those are the ones that the user is intended to access. I followed the explanation that bkpg stated on its webpage about exporting functions, it's possible I understood it incorrectly. It looks like the site is down so I can't check it right now.

I only wanted to allow a usage similar to the first line after the sourcing on your snippet, one for each subcommand turned into command by this modification. However now I see my in my PR the three are functional which is inconsistent and unclear.

>4. I see that you renamed several functions (prefix with `semver_`).  Is this just preference, or, is it an attempt to avoid name collisions within the enclosing (sourcing) script?  If it is collisions, then the problem remains for all the other functions.  Worse, these collisions won't be caught by the shell: functions will silently get re-defined during or after the sourcing.  Here is one advantage of using the current, narrow interface: collisions are not possible. The same problem exists for variables: e.g. imagine that the definition of DIGITS changed!

The reaming of the functions is just for preference, particularly to recognize from what script the functions come from.

>5. You mention the "bpkg" standard.  I assume that you are not proposing making the semver tool a package with a json dependency file, etc.  You simply applied their (neat!) trick of implementing source/invoke in the same script.

I am not proposing making a package out of semver, I just wanted to refer to their sourcing trick.

>6. Error reporting: at present, errors are reported from the perspective of a stand-alone script: e.g. we indicate usage, we claim that the program name is "semver", orient the user to the "-h" option, etc.  Also, `set` is invoked that would change global options for the enclosing script.

This is something I totally overlooked. I'll check into it.

>7. Tests: changes in functionality *must* have automated tests added.

>8. Documentation: the user interface is pretty well documented, including examples.  The same would be needed if sourcing is supported (including caveats about names, errors, ..).

>9. Commit message: if you look at our git log, you will see the sort of commit messages we've produced.  One can understand the evolution of the tool without reverse engineering the code.  "Allow the script to be sourced" is not enough.

Points taken.

> In short, I don't think the PR is ready to go.  If the issues above can be addressed (especially the first, "why?"), then maybe something of reduced scope might work: ditch the exports and renaming, simply add the test for sourcing and replacing `exit` with `return`, add unit tests, and document: "_If you want to, you can source the script.  There are no compatibility guarantees going forward and watch out for the gotcha's._"

I was originally going to submit a issue to request the sourcing of the script as a functionality, however I decided to make a PR instead to express those ideas through code rather than a request.

I am interested in working through the issues you pointed to correctly bring this functionality to semver. If the proposal, not the code which does not suffice, can't be implemented feel free to reject the PR to close the discussion.

**Comment by tvillega** [comment]
I was relocated. Closing.
